v4/weaviate/graphql/multiclassbuilder.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/semi-technologies/weaviate/entities/models"
+)
+
+type MultiClassBuilder struct {
+	connection    rest
+	classBuilders map[string]*builderBase
+}
+
+// ClassName that should be queried
+func NewQueryClassBuilder(className string) *builderBase {
+	return &builderBase{className: className}
+}
+
+func (mb *MultiClassBuilder) AddQueryClass(class *builderBase) *MultiClassBuilder {
+	mb.classBuilders[class.className] = class
+	return mb
+}
+
+// Do execute the GraphQL query
+func (mb *MultiClassBuilder) Do(ctx context.Context) (*models.GraphQLResponse, error) {
+	return runGraphQLQuery(ctx, mb.connection, mb.build())
+}
+
+// build the GraphQL query string (not needed when Do is executed)
+func (mb *MultiClassBuilder) build() string {
+	var query string
+	for _, class := range mb.classBuilders {
+		filterClause := ""
+		if class.includesFilterClause {
+			filterClause = mb.createFilterClause(class)
+		}
+		fieldsClause := mb.createFieldsClause(class)
+		query += fmt.Sprintf("%v %v {%v}", class.className, filterClause, fieldsClause) + " "
+	}
+	query = strings.TrimSpace(query)
+	query = fmt.Sprintf("{Get {%v}}", query)
+	return query
+}
+
+func (mb *MultiClassBuilder) createFilterClause(class *builderBase) string {
+	filters := []string{}
+	if class.withWhereFilter != nil {
+		filters = append(filters, class.withWhereFilter.String())
+	}
+	if class.withNearTextFilter != nil {
+		filters = append(filters, class.withNearTextFilter.build())
+	}
+	if class.withBM25 != nil {
+		filters = append(filters, class.withBM25.build())
+	}
+	if class.withHybrid != nil {
+		filters = append(filters, class.withHybrid.build())
+	}
+	if class.withNearVectorFilter != nil {
+		filters = append(filters, class.withNearVectorFilter.build())
+	}
+	if class.withNearObjectFilter != nil {
+		filters = append(filters, class.withNearObjectFilter.build())
+	}
+	if class.withAskFilter != nil {
+		filters = append(filters, class.withAskFilter.build())
+	}
+	if class.withNearImageFilter != nil {
+		filters = append(filters, class.withNearImageFilter.build())
+	}
+	if class.withGroupFilter != nil {
+		filters = append(filters, class.withGroupFilter.build())
+	}
+	if class.includesLimit {
+		filters = append(filters, fmt.Sprintf("limit: %v", class.limit))
+	}
+	if class.includesOffset {
+		filters = append(filters, fmt.Sprintf("offset: %v", class.offset))
+	}
+	if class.withSort != nil {
+		filters = append(filters, class.withSort.build())
+	}
+	return fmt.Sprintf("(%s)", strings.Join(filters, ", "))
+}
+
+func (mb *MultiClassBuilder) createFieldsClause(class *builderBase) string {
+	if len(class.withFields) > 0 {
+		fields := make([]string, len(class.withFields))
+		for i := range class.withFields {
+			fields[i] = class.withFields[i].build()
+		}
+		return strings.Join(fields, " ")
+	}
+	return ""
+}
+
+// WithFields included in the result set
+func (bb *builderBase) WithFields(fields ...Field) *builderBase {
```

Comment:
Yes you are correct we should avoid defining duplicate methods. But if we remove it from `GetBuilder` and keep it only on `builderBase` then it would be a breaking change. So on the `GetBuilder` structure we can only change underlying structure not the methods defined on it. 

We need to define these methods on `builderBase` because in `MultiClassBuilder` structure, underlying we are using `builderBase`.

Reply:
I had a bit more time today and played a bit around with your code.
Thanks to your explanation above I made a few changes that reduce the amount of duplicate code, have a look at this branch: https://github.com/semi-technologies/weaviate-go-client/tree/pr/106.

What do you think?

Reply:
I had similar approach in mind but I wanted to keep `MultiClassBuilder` loosely coupled with `GetBuilder`. Doing this is exactly opposite of that, and few other reasons are as:

1. Earlier there was only one way to get `GetBuilder` but now since `MultiClassBuilder` also refers to `GetBuilder`, user can do something like this : 
```
builder := graphql.NewQueryClassBuilder("Pizza")
```
and it will return the variable of type `*graphql.GetBuilder` and on this `Do()` method is also defined. Even though calling it will only return error, but it makes it a bit confusing.

2. Now in `MultiClassBuilder` struct we have multiple connection available. One in `MultiClassBuilder` itself and every time we add a new class to `MultiClassBuilder`, that class will have one connection available. This won't harm in anyway, but it doesn't make sense and it's just not good design.

We can go ahead with this approach but instead of solving legacy design problems, we'll just be adding more bad design and in future it while trying to solve this, it'll become breaking changes. So I would prefer having some duplication over adding confusing design.

Do let me know your thoughts.

Reply:
Sorry for the late reply, we're in the middle of renaming the company and that has kept me busy the last days.

I see the downsides of my approach, but I'd go for easier mantainability when there is a choice. I'm not happy with copying ~300 lines of code and I'm sure that one of them will be forgotten when there is the next addition/fix to the builder(s). 

As a compromise:
- You can move createFilterClause + createFieldsClause to builderBase and use them without problems from both implementations (do class.createFieldsClause() instead of mb.createFieldsClause(class)). That would cut the amount copied code by ~200 lines. 
- Moreover, we would need tests for each of the remaining duplicated functions + integration tests.

Does that work for you? @antas-marcin what do you think?

Reply:
I've made the suggested changes, let me know if anymore changes are required.

Reply:
Cool, thanks!

I think an integration test would be great, as far as I see the "Do" function is also not used yet. And maybe an unittest with more than one property? Then we should have good test coverage and can merge afterwards :)

Reply:
Would be great if you could rebase onto the latest main. I think I fixed the CI for PRs from forks

Reply:
I've re-based it with current main and also I've added few unit tests, let me know if this works or not after that will start working on adding integration test.

Reply:
Tests run, thanks!

A few [seem to be failing](https://github.com/weaviate/weaviate-go-client/pull/106/commits/7a10e9c20c10bb6f95169b60cf8e0442b302acda), could you have a look? I can also have a look on Monday

Reply:
I just saw that, test : `TestMultiClassQueryBuilder/Multiple_filter#01`, is an error from my end, I'll fix that. But test : `TestMultiClassBM25Builder` looks like an pure ordering issue. Should we sort based on `className` while building the query?

Reply:
> Should we sort based on className while building the query?

That should be pretty cheap, so yes let's do that!

Reply:
Hey, @dirkkul Sorry for the delay. Got busy with some priority work. I've fixed the unit test issue. While working on integration tests for `MultiClassBuilder`, got to know that these two `TestGraphQL_integration/Get_bm25_filter`, `TestGraphQL_integration/Get_hybrid_filter` integration tests cases are failing. It's the same way in main branch also. 

It would be great if you can quickly look into it and share your thoughts. Meanwhile I will add other integration tests for `MultiClassBuilder`.

Reply:
No worries, take your time :)

I added a comment, thanks for updating!